File: Data Structures & Algorithms/topologicalSort/submission-0.py
from collections import defaultdict

class Solution:
    def topologicalSort(self, n: int, edges: List[List[int]]) -> List[int]:
        adj = defaultdict(list)
        vis = {i:0 for i in range(n)}

        for e in edges:
            adj[e[0]].append(e[1])

        res=[]
        def iscycle(node):
            if vis[node] == 1:
                return True
            if vis[node] == 2:
                return False
            
            vis[node] = 1
            for x in adj[node]:
                if iscycle(x):
                    return True
            
            vis[node] = 2
            res.append(node)
            return False
            
        for i in range(n):
            if vis[i]==0 and iscycle(i):
                return []

        return res[::-1]

File: Data Structures & Algorithms/count-connected-components/submission-0.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        unordered_map<int, vector<int>> adj;
        for(auto &e: edges){
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        set<int> vis;

        int res=0;
        for(int i=0;i<n;i++){
            if(!vis.count(i)){
                dfs(adj, vis, i);
                res++;
            }
        }
        return res;
    }

    void dfs(unordered_map<int, vector<int>> &adj, set<int> &vis, int i){
        if(vis.count(i)){
            return;
        }
        vis.insert(i);
        for(auto &n: adj[i]){
            dfs(adj, vis, n);
        }
    }
};
